// simple-body-validator/lib/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setTranslationObject = exports.setTranslationPath = exports.setDefaultLang = exports.make = exports.Password = exports.ImplicitRule = exports.Rule = void 0;
var validator_1 = __importDefault(require("./validator"));
var lang_1 = __importDefault(require("./lang"));
var ruleContract_1 = __importDefault(require("./rules/ruleContract"));
var implicitRuleContract_1 = __importDefault(require("./rules/implicitRuleContract"));
var password_1 = __importDefault(require("./rules/password"));
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Rule;
}(ruleContract_1.default));
exports.Rule = Rule;
;
var ImplicitRule = /** @class */ (function (_super) {
    __extends(ImplicitRule, _super);
    function ImplicitRule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ImplicitRule;
}(implicitRuleContract_1.default));
exports.ImplicitRule = ImplicitRule;
;
var Password = /** @class */ (function (_super) {
    __extends(Password, _super);
    function Password() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Password;
}(password_1.default));
exports.Password = Password;
;
__exportStar(require("./rules/registerRule"), exports);
__exportStar(require("./rule"), exports);
function make(data, rules, customMessages) {
    if (data === void 0) { data = {}; }
    if (rules === void 0) { rules = {}; }
    if (customMessages === void 0) { customMessages = {}; }
    return new validator_1.default(data, rules, customMessages);
}
exports.make = make;
;
function setDefaultLang(lang) {
    lang_1.default.setDefaultLang(lang);
}
exports.setDefaultLang = setDefaultLang;
;
function setTranslationPath(path) {
    lang_1.default.setPath(path);
}
exports.setTranslationPath = setTranslationPath;
;
function setTranslationObject(translations) {
    lang_1.default.setTranslationObject(translations);
}
exports.setTranslationObject = setTranslationObject;
