// simple-body-validator/lib/rules/ruleContract.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var lang_1 = __importDefault(require("../lang"));
var object_1 = require("../utils/object");
var RuleContract = /** @class */ (function () {
    function RuleContract() {
        /**
         * The validation error message.
         */
        this.message = '';
        /**
         * All of the data under validation.
         */
        this.data = {};
    }
    /**
     *  Determine if the validation rule passes.
     */
    RuleContract.prototype.passes = function (value, attribute) {
        return true;
    };
    ;
    /**
     * Get the validation error message.
     */
    RuleContract.prototype.getMessage = function () {
        return this.message;
    };
    ;
    /**
     * Set the data under validation.
     */
    RuleContract.prototype.setData = function (data) {
        this.data = data;
        return this;
    };
    ;
    /**
     * Set the tranlation language
     */
    RuleContract.prototype.setLang = function (lang) {
        this.lang = lang;
        return this;
    };
    ;
    /**
     * Get the translated error message based on the specified path
     */
    RuleContract.prototype.trans = function (path, params) {
        if (params === void 0) { params = {}; }
        var validatonMessages = lang_1.default.get(this.lang);
        var message = (0, object_1.deepFind)(validatonMessages, path) || '';
        if (!message) {
            return message;
        }
        for (var key in params) {
            message = message.replace(":".concat(key), params[key]);
        }
        return message;
    };
    return RuleContract;
}());
exports.default = RuleContract;
