// simple-body-validator/lib/validators/replaceAttributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_1 = require("../utils/date");
var formatMessages_1 = require("../utils/formatMessages");
var general_1 = require("../utils/general");
var object_1 = require("../utils/object");
var replaceAttributes = {
    /**
     * Replace all place-holders for the accepted_if rule.
     */
    replaceAcceptedIf: function (message, parameters, data) {
        var other = parameters[0];
        var result = (0, object_1.deepFind)(data, other);
        var values = {
            ':other': (0, formatMessages_1.getDisplayableAttribute)(other),
            ':value': result
        };
        return message.replace(/:other|:value/gi, function (matched) { return values[matched]; });
    },
    /**
     * Replace all place-holders for the after rule.
     */
    replaceAfter: function (message, parameters) {
        return this.replaceBefore(message, parameters);
    },
    /**
     * Replace all place-holders for the after_or_equal rule.
     */
    replaceAfterOrEqual: function (message, parameters) {
        return this.replaceBefore(message, parameters);
    },
    /**
     *  Replace all place-holders for the before rule.
     */
    replaceBefore: function (message, paramaters) {
        if (!(0, date_1.toDate)(paramaters[0])) {
            return message.replace(':date', (0, formatMessages_1.getDisplayableAttribute)(paramaters[0]));
        }
        return message.replace(':date', paramaters[0]);
    },
    /**
     * Replace all place-holders for the before_or_equal rule.
     */
    replaceBeforeOrEqual: function (message, parameters) {
        return this.replaceBefore(message, parameters);
    },
    /**
     * Replace all the place-holders for the between rule.
     */
    replaceBetween: function (message, parameters) {
        var values = { ':min': parameters[0], ':max': parameters[1] };
        return message.replace(/:min|:max/gi, function (matched) { return values[matched]; });
    },
    /**
     * Replace all place-holders for the before_or_equal rule.
     */
    replaceDateEquals: function (message, parameters) {
        return this.replaceBefore(message, parameters);
    },
    /**
     *  Replace all place-holders for the declined_if rule.
     */
    replaceDeclinedIf: function (message, parameters, data) {
        var other = parameters[0];
        var result = (0, object_1.deepFind)(data, other);
        var values = {
            ':other': (0, formatMessages_1.getDisplayableAttribute)(other),
            ':value': result
        };
        return message.replace(/:other|:value/gi, function (matched) { return values[matched]; });
    },
    /**
     * Replace all place-holders for the different rule.
     */
    replaceDifferent: function (message, parameters) {
        return this.replaceSame(message, parameters);
    },
    /**
     * Replace all place-holders for the digits rule.
     */
    replaceDigits: function (message, parameters) {
        return message.replace(':digits', parameters[0]);
    },
    /**
     * Replace all place-holders for the digits (between) rule.
     */
    replaceDigitsBetween: function (message, parameters) {
        return this.replaceBetween(message, parameters);
    },
    /**
     * Replace all place-holders for the ends_with rule.
     */
    replaceEndsWith: function (message, parameters) {
        return message.replace(':values', parameters.join(', '));
    },
    /**
     * Replace all place-holders for the starts_with rule.
     */
    replaceStartsWith: function (message, parameters) {
        return message.replace(':values', parameters.join(', '));
    },
    /**
     * Replace all place-holders for the min rule.
     */
    replaceMin: function (message, parameters) {
        return message.replace(':min', parameters[0]);
    },
    /**
     * Replace all place-holders for the max rule.
     */
    replaceMax: function (message, parameters) {
        return message.replace(':max', parameters[0]);
    },
    /**
     * Replace all place-holders for the required_with rule.
     */
    replaceRequiredWith: function (message, parameters) {
        return message.replace(':values', parameters.map(function (attribute) { return (0, formatMessages_1.getDisplayableAttribute)(attribute); }).join(', '));
    },
    /**
     * Replace all place-holders for the required_with_all rule.
     */
    replaceRequiredWithAll: function (message, parameters) {
        return this.replaceRequiredWith(message, parameters);
    },
    /**
     * Replace all place-holders for the required_without rule.
     */
    replaceRequiredWithout: function (message, parameters) {
        return this.replaceRequiredWith(message, parameters);
    },
    /**
     * Replace all place-holders for the required_without_all rule.
     */
    replaceRequiredWithoutAll: function (message, parameters) {
        return this.replaceRequiredWith(message, parameters);
    },
    /**
     * Replace all place-holders for the gt rule.
     */
    replaceGt: function (message, parameters, data, hasNumericRule) {
        var value = parameters[0];
        var result = (0, object_1.deepFind)(data, value);
        if (typeof result === 'undefined') {
            return message.replace(':value', value);
        }
        return message.replace(':value', (0, general_1.getSize)(result, hasNumericRule).toString());
    },
    /**
     * Replace all place-holders for the lt rule.
     */
    replaceLt: function (message, parameters, data, hasNumericRule) {
        return this.replaceGt(message, parameters, data, hasNumericRule);
    },
    /**
    * Replace all place-holders for the gte rule.
    */
    replaceGte: function (message, parameters, data, hasNumericRule) {
        return this.replaceGt(message, parameters, data, hasNumericRule);
    },
    /**
     * Replace all place-holders for the lte rule.
     */
    replaceLte: function (message, parameters, data, hasNumericRule) {
        return this.replaceGt(message, parameters, data, hasNumericRule);
    },
    /**
     * Replace all place-holders for the required_if rule.
     */
    replaceRequiredIf: function (message, parameters, data) {
        var other = parameters[0];
        var result = (0, object_1.deepFind)(data, other);
        var values = {
            ':other': (0, formatMessages_1.getDisplayableAttribute)(other),
            ':value': result
        };
        return message.replace(/:other|:value/gi, function (matched) { return values[matched]; });
    },
    /**
     * Replace all the place-holders for the required_unless rule.
     */
    replaceRequiredUnless: function (message, parameters) {
        var other = parameters[0];
        var values = {
            ':other': (0, formatMessages_1.getDisplayableAttribute)(other),
            ':values': parameters.slice(1).join(', ')
        };
        return message.replace(/:other|:values/gi, function (matched) { return values[matched]; });
    },
    /**
     * Replace all place-holders for the same rule.
     */
    replaceSame: function (message, parameters) {
        return message.replace(':other', (0, formatMessages_1.getDisplayableAttribute)(parameters[0]));
    },
    /**
     * Replace all place-holders for the size rule.
     */
    replaceSize: function (message, parameters) {
        return message.replace(':size', parameters[0]);
    },
};
exports.default = replaceAttributes;
